### ISSUE TITLE
Validator client hangs if permissions are not set correctly

### ISSUE THREAD

**Issue by michaelsproul** [sigp/lighthouse#2430]
## Description

The VC sometimes hangs (deadlocks) in a mysterious way when permissions aren't set correctly on its files. There are likely multiple permutations of this bug, based on exactly which files or directories are inaccessible: we've seen several users report hangs at different points.

The hang that I've been experimenting is involves setting the permissions on `$datadir/validator` so that it can't be read, i.e.

```
$ ls -lah ~/.lighthouse/pyrmont/validators
total 8.0K
drwxr-xr-x 2 root root 4.0K Jul  5 12:51 ./
drwxrwxr-x 3 root root 4.0K Jul  5 12:51 ../
```

Starting the validator client in this circumstance yields:

```
$ lighthouse vc --network pyrmont
Jul 05 16:09:49.102 INFO Lighthouse started                      version: Lighthouse/v1.4.0-3b600ac
Jul 05 16:09:49.102 INFO Configured for network                  name: pyrmont
Jul 05 16:09:49.102 INFO Starting validator client               validator_dir: "/home/michael/.lighthouse/pyrmont/validators", beacon_nodes: ["http://localhost:5052/"]
Jul 05 16:09:49.102 INFO HTTP metrics server is disabled
```

No further output is produced.

I've isolated the call that blocks as `File::create` here: https://github.com/sigp/lighthouse/blob/3b600acdc5bf9726367c18277a22486573b8b457/common/filesystem/src/lib.rs#L58 Replacing that call by `tokio::fs::File::create` didn't help. `File::create` also doesn't hang if you call it on the same file _off_ the Tokio executor.

Running under GDB shows that all threads end up parked waiting on each other. The backtraces mostly look like this:

```
#0  syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
#1  0x0000555556ba6f70 in parking_lot_core::thread_parker::imp::ThreadParker::futex_wait (self=0x7fff9e5f2158, ts=...)
    at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/parking_lot_core-0.8.3/src/thread_parker/linux.rs:112
#2  parking_lot_core::thread_parker::imp::{{impl}}::park_until (self=0x7fff9e5f2158, timeout=...) at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/parking_lot_core-0.8.3/src/thread_parker/linux.rs:87
#3  parking_lot_core::parking_lot::park::{{closure}}<closure-0,closure-1,closure-2> (thread_data=0x7fff9e5f2138) at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/parking_lot_core-0.8.3/src/parking_lot.rs:609
#4  parking_lot_core::parking_lot::with_thread_data<parking_lot_core::parking_lot::ParkResult,closure-0> (f=...) at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/parking_lot_core-0.8.3/src/parking_lot.rs:183
#5  parking_lot_core::parking_lot::park<closure-0,closure-1,closure-2> (key=<optimised out>, validate=..., before_sleep=..., timed_out=..., park_token=..., timeout=...)
    at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/parking_lot_core-0.8.3/src/parking_lot.rs:576
#6  parking_lot::condvar::Condvar::wait_until_internal (self=<optimised out>, mutex=<optimised out>, timeout=...) at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/parking_lot-0.11.1/src/condvar.rs:338
#7  0x0000555556753555 in parking_lot::condvar::Condvar::wait_for<tokio::runtime::blocking::pool::Shared> (self=0x55555b41e1d0, mutex_guard=0x7fff9e5f17d0, timeout=...)
    at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/parking_lot-0.11.1/src/condvar.rs:397
#8  tokio::loom::std::parking_lot::Condvar::wait_timeout<tokio::runtime::blocking::pool::Shared> (self=0x55555b41e1d0, guard=..., timeout=...)
    at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.7.1/src/loom/std/parking_lot.rs:100
#9  tokio::runtime::blocking::pool::Inner::run (self=0x55555b41e130, worker_thread_id=<optimised out>) at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.7.1/src/runtime/blocking/pool.rs:274
#10 tokio::runtime::blocking::pool::{{impl}}::spawn_thread::{{closure}} () at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.7.1/src/runtime/blocking/pool.rs:245
#11 std::sys_common::backtrace::__rust_begin_short_backtrace<closure-0,()> (f=...) at /home/michael/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:125
...
```

I.e. it looks like a deadlock internal to Tokio itself (using Tokio's `parking_lot_core`, _not_ `parking_lot`).

## Version

Lighthouse v1.4.0

## Steps to resolve

* Sift through the backtraces for all the threads in `gdb` to see if one of them shows any useful information. I've spot checked the 36 threads on my machine and they all show output similar to the above.
* Try switching all the locks in the validator client to async-aware ones. This is a bit like blindly trying to hammer a nail in the dark, but it might work and I'm out of other ideas.

**Comment by michaelsproul** [comment]
Another option: mitigate this by checking the file permissions in `ensure_dir_exists` or something. Checking earlier (or not relying on `File::create`) might allow us to sidestep the issue.

**Comment by paulhauner** [comment]
Looking at the logs, it seems likely we're in the following call stack:

1. [`create_with_600_perms`](https://github.com/sigp/lighthouse/blob/3b600acdc5bf9726367c18277a22486573b8b457/common/filesystem/src/lib.rs#L58)
1. [`write_file_via_temporary`](https://github.com/sigp/lighthouse/blob/3b600acdc5bf9726367c18277a22486573b8b457/common/account_utils/src/lib.rs#L68)
1. [`ValidatorDefinitions::save`](https://github.com/sigp/lighthouse/blob/3b600acdc5bf9726367c18277a22486573b8b457/common/account_utils/src/validator_definitions.rs#L262)
1. [`ValidatorDefinitions::open_or_create`](https://github.com/sigp/lighthouse/blob/3b600acdc5bf9726367c18277a22486573b8b457/common/account_utils/src/validator_definitions.rs#L121)
1. [`ProductionValidatorClient::new_from_cli`](https://github.com/sigp/lighthouse/blob/3b600acdc5bf9726367c18277a22486573b8b457/validator_client/src/lib.rs#L79)

If that's the case, we'd only be running a single Tokio task (the one that eventually spawns the other tasks). That would mean that the `File::create` call is doing all the blocking and the deadlock isn't caused by interaction between our tasks.

I would be interesting to call this on your example filesystem in a non-Tokio, single-threaded context and see if it still blocks:

https://github.com/sigp/lighthouse/blob/3b600acdc5bf9726367c18277a22486573b8b457/validator_client/src/lib.rs#L141-L142

**Comment by michaelsproul** [comment]
Yeah, that's the code path, but it seemed like there were multiple Tokio tasks running already when I was looking with GDB.

I also realised that #2394 is pretty much an instance of this same problem (bad permissions causing a hang). But I'll leave it open because it's specific to how we set perms on Windows.

**Comment by michaelsproul** [comment]
Another way to trigger a hang: change a path in `validator_definitions.yml` so that it points to a file that doesn't exist, or equivalently move your validators dir without munging the paths.

**Comment by paulhauner** [comment]
Perhaps this could be the culprit:

https://github.com/sigp/lighthouse/blob/3b600acdc5bf9726367c18277a22486573b8b457/lighthouse/src/main.rs#L377

If we get an error during `ProductionValidatorClient::new`, we fail to log an error or shutdown the client. Instead, we just return an error into the abyss of a `Runtime::spawn`.

**Comment by paulhauner** [comment]
I believed this was solved in #2436, feel free to reopen if you disagree :)
